### PR TITLE
Add "not tool" function to extended drop tables, document how extended drop tables work.

### PR DESCRIPTION
**Pull request by ashtrayoz**
This functionality has apparently been around for years, but undocumented.
I have tried to follow the most common whitespace conventions in the file, but it is inconsistent, so any review comments on this would be helpful.

### CODE REVIEW

**Comment by ashtrayoz** [comment]
Pull request #7224 fixes a bug related to this.
This pull request can be merged either before or after 7224.
So if the implementation of 7224 changes, it should not delay this.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -5394,6 +5395,37 @@ Definition tables
                 },
             },
         },
+        ^ This can be filtered by tool type, so that different tools
+        ^ cause different drops, or so that particular tools do not
+        ^ cause particular drops:
+        minetest.register_node("stonesand:stone_sand", { description = "Stone with sand",
+            tiles = { "default_sand.png", "default_stone.png", "default_sand.png^[lowpart:50:default_stone.png" },
+            groups = { cracky=3, crumbly=3 },
```
Comment (SmallJoker) [inline doc/lua_api.txt]: `drop` table only. This definition data is irrelevant and not helpful.

File: doc/lua_api.txt [hunk 2]
```diff
@@ -5394,6 +5395,37 @@ Definition tables
...
+                    },
+                    {
+                        tools = { "default:shovel_wood", "default:shovel_stone", "default:shovel_steel",
+                            "default:shovel_bronze", "default:shovel_mese", "default:shovel_diamond" },
+                        rarity = 5,         -- Less common.
+                        items = { "default:desert_sand" },
+                    },
+                    -- If they are not using a shovel, they will always get a drop.
+                    {
+                        tools = { "~default:shovel" },
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Needs explanation what exactly `~` does.

**Comment by SmallJoker** [comment]
IMO you can combine both PRs. Features need documentation - and documentation without the according feature is wrong. Maybe it would look a bit different if you changed more than a thousand lines.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -5380,20 +5380,37 @@ Definition tables
             place = <SimpleSoundSpec>,
             place_failed = <SimpleSoundSpec>,
         },
-        drop = "",
+
+        drop = "", --[[
         ^ Name of dropped node when dug. Default is the node itself.
-        ^ Alternatively:
+        ^ Using a table allows multiple items, random selection among items,
+        ^ palette control and tool filtering:
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Using `^` here confuses. Only spaces only for these two lines.

File: doc/lua_api.txt [hunk 4]
```diff
@@ -5380,20 +5380,37 @@ Definition tables
...
+                    rarity = 2,  -- Probability of dropping is 1 / rarity - default:sand is common.
+                    items = { "default:sand" },  -- Items to drop.
+                },
+                {
+                    tools = { "default:shovel_wood", "default:shovel_stone", "default:shovel_steel",
+                        "default:shovel_bronze", "default:shovel_mese", "default:shovel_diamond" },
+                    rarity = 5,         -- default:desert_sand is less common.
+                    items = { "default:desert_sand" },
+                },
+                -- "~" means "a tool not matching the following expression" - string.find is used for the match.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: This line is too long (>80 characters), please break it.

File: doc/lua_api.txt [hunk 5]
```diff
@@ -5380,20 +5380,37 @@ Definition tables
...
+                -- If they are using a shovel, they will get a drop about 60% of the time.
+                {
+                    tools = { "default:shovel_wood", "default:shovel_stone", "default:shovel_steel",
+                        "default:shovel_bronze", "default:shovel_mese", "default:shovel_diamond" },
+                    rarity = 2,  -- Probability of dropping is 1 / rarity - default:sand is common.
+                    items = { "default:sand" },  -- Items to drop.
+                },
+                {
+                    tools = { "default:shovel_wood", "default:shovel_stone", "default:shovel_steel",
+                        "default:shovel_bronze", "default:shovel_mese", "default:shovel_diamond" },
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Keep examples simple. Two or three items are already clear enough.

**Comment by SmallJoker** [comment]
Thanks, the documentation looks good now. However, it will document the behaviour in a wrong way:
`"~" means "a tool not matching the following expression"`. Currently this is not true and should be `a tool matching the following expression`, unless you'd include the PR #7224 here (which looks good to me). Please either include the PR or correct the documentation so it describes the current behaviour.

**Comment by ashtrayoz** [comment]
I think I understand more use cases now, which relates to the last comment in #7224.  Apologies if the following is too obvious or drawn out, but I think it is worth getting right, and I am not sure that I have done that yet.

In my world, this code will be used for about 30 or 40 nodes (and growing), and for a class of tools which does not exist in minetest_game.  But for the purposes of discussion, we can talk about using a shovel (1 of the 6 types in minetest_game) on the node "stone_with_sand".

In natural language, what a modder might want is "drop sand if dug with a shovel, and drop cobble if dug with anything else".  We need Lua expressions for the noun phrases "a shovel" and "anything else", which is equivalent to "not a shovel".

In some languages, "not a shovel" could be written as "^(?!default:shovel_).*", but I am not aware of any such construct in Lua.  But in Lua, "not equal" is "~=".  Which suggests that the "~" in the current code *could* mean "*not* the following pattern".  This is what #7224 implemented.

But!  What is the best way to express "a shovel"?  In this pull request, I have expressed it as 
    {
    "default:shovel_wood", "default:shovel_stone",
    "default:shovel_bronze", "default:shovel_steel",
    "default:shovel_mese", "default:shovel_diamond",
    }
which is rather long-winded.  Perhaps we need a way of saying "an itemname which matches the following expression when passed to string.find".  If we choose an arbitrary magic character as a flag for this behavious, say "^", then the five lines of code above can be shortened to "^default:shovel_".

So there are three ways that strings could be used to name tools:
1. literal match
1. positive pattern match
1. negative pattern match

The code in master currently implements two of these.  The code in #7224 implements a different two of these.  I now think that we should implement all three.  We could even add a fourth if we wished:  negative literal match.

We get to choose syntax.  In 7224, I reused existing syntax, but this could break existing mods/worlds - if there are any mods/worlds out there that use extended drop tables with tool names and the "~" operator.  It is hard to know how many of these exist - they are certainly not common - but it is probably good policy to avoid breaking them if possible.  We can also decide whether we want the syntax to be orthogonal:  should "pattern" and "not" be separate items of syntax, or should "not this pattern" be expressed by a single flag which conflates the two ideas?

If "normal:string" is any string which has no magic character, one possible syntax is:
  - "normal:string" - match literally
  - "~normal:string" - match as a pattern
  - "^normal:string" - negative match literally
  - "^~normal:string" - negative match as a pattern

Another possible syntax is:
  - "normal:string" - match literally
  - "~normal:string" - match as a pattern
  - "!normal:string" - negative match literally
  - "?normal:string" - negative match as a pattern

This would lead to an extended drop table like:
    drop = {
        max_items = 1,
        items = {
            {
            tools = { "~default:shovel_" },
            items = { "default:sand" },
            },
            {
            tools = { "?default:shovel_" },
            items = { "default:cobble" },
            }
        }
    }

What do other people think?

File: doc/lua_api.txt [hunk 6]
```diff
@@ -5380,20 +5380,49 @@ Definition tables
...
+
+        drop = "", --[[
         ^ Name of dropped node when dug. Default is the node itself.
-        ^ Alternatively:
+          Using a table allows multiple items, random selection among items,
+          palette control and tool filtering:
         drop = {
             max_items = 1,  -- Maximum number of items to drop.
             items = {  -- Choose max_items randomly from this list.
+                -- When using a shovel, they will get a drop about 60% of the time.
```
Comment (Ezhh) [inline doc/lua_api.txt]: Who is "they"? Please be specific ("..player will get...")

File: doc/lua_api.txt [hunk 7]
```diff
@@ -5380,20 +5380,49 @@ Definition tables
...
+                {
+                    -- Tool names are matched with `==`.
+                    tools = {
+                        "default:shovel_wood", "default:shovel_stone",
+                        "default:shovel_bronze", "default:shovel_steel",
+                        "default:shovel_mese", "default:shovel_diamond",
+                    },
+                    rarity = 2,  -- Probability of dropping is 1 / rarity.
```
Comment (Ezhh) [inline doc/lua_api.txt]: Is this a separate 1/rarity check for each possible item, as defined in max_items? Couldn't hurt to explain.

File: doc/lua_api.txt [hunk 8]
```diff
@@ -5380,20 +5380,49 @@ Definition tables
...
-        drop = "",
+
+        drop = "", --[[
         ^ Name of dropped node when dug. Default is the node itself.
-        ^ Alternatively:
+          Using a table allows multiple items, random selection among items,
+          palette control and tool filtering:
         drop = {
             max_items = 1,  -- Maximum number of items to drop.
             items = {  -- Choose max_items randomly from this list.
```
Comment (Ezhh) [inline doc/lua_api.txt]: What list? The list of items seems to come way after this comment, so there's some space for confusion.

File: doc/lua_api.txt [hunk 9]
```diff
@@ -5380,20 +5380,49 @@ Definition tables
...
+                        "default:shovel_wood", "default:shovel_stone",
+                        "default:shovel_bronze", "default:shovel_steel",
+                        "default:shovel_mese", "default:shovel_diamond",
+                    },
+                    rarity = 2,  -- Probability of dropping is 1 / rarity.
+                    items = { "default:sand" },  -- Items to drop.
+                },
+                {
+                    -- The next entry uses "~", so this list should be its exact
+                    -- opposite.  If this list is not exhaustive, there are extra
```
Comment (Ezhh) [inline doc/lua_api.txt]: The next entry appears to be lines 5406 to 5413, which is identical to the previous one apart from seeming to show a rarity value change for a different node. The part with ~ doesn't start until 5419 and seems to have a separate comment introducing it. This should probably be structured better, with less repetition and comments more tightly linked to what they reference.

File: doc/lua_api.txt [hunk 10]
```diff
@@ -5380,20 +5380,49 @@ Definition tables
...
+                        "default:shovel_wood", "default:shovel_stone",
+                        "default:shovel_bronze", "default:shovel_steel",
+                        "default:shovel_mese", "default:shovel_diamond",
+                    },
+                    rarity = 5,  -- default:desert_sand is less common.
+                    items = { "default:desert_sand" },
+                },
+                -- "~" means "a tool not matching the following expression"
+                -- `string.find` is used for the match.
+                -- So if they are not using some sort of shovel, they will always
```
Comment (Ezhh) [inline doc/lua_api.txt]: Please use "the player", not "they".

**Comment by Wuzzy2** [comment]
Thanks for adding the documentation. Wow, I was not aware you can actually test which tool was used to dig the node. This changes everything! :D

---

But this also reveals some really odd features in Minetest. I mean, string matching for items? WTF? Do we even need this string matching nowadays? This feature seems pretty pointless and overcomplicated because we have groups. I believe there should be no string matching or regex or anything like that. Just use groups. Too bad that Minetest Game does not have proper tool groups (like `shovel=1`, or `pickaxe=1`) yet, but that's a different story. String matching will cause all sorts of “funny” surprises when a mod update adds a tool which unexpectedly matches your regex, but you actually don't wanted that particular tool.
Could it be that string matching has been introduced before groups were a thing?

As for the question mark, this should be only used as a logical NOT for an exact itemstring, not a regular expression, i.e. `?default:shovel_wood` = “anything EXCEPT `default:shovel_wood`”.

I suggest to remove all string matching capabilities in a different PR first.

**Comment by SmallJoker** [comment]
I like how the documentation currently is - changes in this feature belong to another PR IMO.
@ashtrayoz Could you please rebase this PR so it can be merged?

**Comment by paramat** [comment]
@SmallJoker is the documentation here mergeable? If so i can rebase it and merge this. No response from author.

**Comment by SmallJoker** [comment]
@paramat Yes they are. Feel free to rebase & merge as soon you've got time for it.

**Comment by paramat** [comment]
Will do.

**Comment by paramat** [comment]
Replaced by #8458

**Comment by paramat** [comment]
https://github.com/minetest/minetest/blob/0cc85a7dc6d7840c47dbd9a9193f5eaaee10dad3/builtin/game/item.lua#L219

I can't see any detection of '?' in the code.
This PR seems to contain documentation of a non-existent feature. Will fix in my PR.

**Pull request closed without merging**